src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -373,5 +378,131 @@ private static void TrimInaccessibleText(ArrayBuilder<SourceText> segments)
+            {
+                _compositeText = compositeText;
+                _segmentLineNumbers = new int[compositeText.Segments.Length];
+
+                var accumulatedLineCount = 0;
+                for (int i = 0; i < compositeText.Segments.Length; i++)
+                {
+                    _segmentLineNumbers[i] = accumulatedLineCount;
+
+                    var segment = compositeText.Segments[i];
+                    accumulatedLineCount += (segment.Lines.Count - 1);
+
+                    // If "\r\n" is split amongst adjacent segments, reduce the line count by one as both segments
+                    // would count their corresponding CR or LF as a newline.
+                    if (segment.Length > 0 &&
+                        segment[^1] == '\r' &&
+                        i < compositeText.Segments.Length - 1)
+                    {
+                        var nextSegment = compositeText.Segments[i + 1];
+                        if (nextSegment.Length > 0 && nextSegment[0] == '\n')
+                        {
+                            accumulatedLineCount -= 1;
+                        }
+                    }
+                }
+
+                _lineCount = accumulatedLineCount + 1;
+            }
+
+            public override int Count => _lineCount;
+
+            public override int IndexOf(int position)
+            {
+                _compositeText.GetIndexAndOffset(position, out var segmentIndex, out var segmentOffset);
+
+                var segment = _compositeText.Segments[segmentIndex];
+                var lineNumberWithinSegment = segment.Lines.IndexOf(segmentOffset);
+
+                return _segmentLineNumbers[segmentIndex] + lineNumberWithinSegment;
+            }
+
+            public override TextLine this[int lineNumber]
+            {
+                get
+                {
+                    // Determine the indices for segments that contribute to our view of this line's contents
+                    GetSegmentIndexRangeContainingLine(lineNumber, out var firstSegmentIndex, out var lastSegmentIndex);
+
+                    var firstSegmentFirstLineNumber = _segmentLineNumbers[firstSegmentIndex];
+                    var firstSegment = _compositeText.Segments[firstSegmentIndex];
+                    var firstSegmentOffset = _compositeText._segmentOffsets[firstSegmentIndex];
+                    var firstSegmentTextLine = firstSegment.Lines[lineNumber - firstSegmentFirstLineNumber];
+
+                    var lineLength = firstSegmentTextLine.SpanIncludingLineBreak.Length;
+
+                    // walk forward through remaining segments contributing to this line, and add their
+                    // view of this line.
+                    for (var nextSegmentIndex = firstSegmentIndex + 1; nextSegmentIndex <= lastSegmentIndex; nextSegmentIndex++)
+                    {
+                        var nextSegment = _compositeText.Segments[nextSegmentIndex];
```

Comment:
add asserts for the middle ones.  put comment that hte last one has at least one.  but then that even if it has many, all we care about it the first, since that't he one that has the new line, which woudl be the newline for the composite's line